pkg/deployment/agency/state.go

```diff
@@ -115,3 +119,72 @@ func (d *StateExists) UnmarshalJSON(bytes []byte) error {
 	*d = bytes != nil
 	return nil
 }
+
+type CollectionIterator func(db, col string, planCollection *StatePlanCollection, shard string, plan ShardServers, current ShardServers)
+
+func (s State) IterateOverCollections(i CollectionIterator) {
+	for db, collections := range s.Plan.Collections {
+		for collection, details := range collections {
+			for shard, shardDetails := range details.Shards {
+				if currShard, ok := s.Current.Collections[db][collection][shard]; ok {
+					i(db, collection, &details, shard, shardDetails, currShard.Servers)
+				}
+			}
+		}
+	}
+}
+
+func (s State) IsDBServerInSync(serverID string) bool {
+	isInSync := true
+	s.IterateOverCollections(func(db, col string, planCollection *StatePlanCollection, shard string, plan ShardServers, current ShardServers) {
+		if !isInSync {
+			return
+		}
+		serverIsNotInSync := plan.Contains(serverID) && !current.Contains(serverID)
+		if serverIsNotInSync {
+			isInSync = false
+			return
+		}
+	})
+
+	return isInSync
+}
+
+func (s State) IsDBServerReadyToRestart(serverID string) bool {
+	readyToRestart := true
+
+	s.IterateOverCollections(func(db, col string, planCollection *StatePlanCollection, shard string, plan ShardServers, current ShardServers) {
```

Comment:
maybe the function which is provided as an argument should return `bool` so we can stop iterating over shards earlier?

Reply:
done